waiter/src/waiter/scheduler.clj

```diff
@@ -638,3 +643,121 @@
           "WAITER_PASSWORD" (service-id->password-fn service-id)
           "WAITER_SERVICE_ID" service-id
           "WAITER_USERNAME" "waiter"}))
+
+;; Support for tracking service instance launching time stats
+
+(defrecord ServiceInstancesLaunchTracker [service-id known-instance-ids scheduling-instance-timer-contexts starting-instance-context-maps])
+
+(defn- new-tracker-val
+  [service-id service-id->service-description-fn]
+  (let [{:strs [min-instances]} (service-id->service-description-fn service-id)
+        ctxs (repeatedly
+               min-instances
+               (fn [] {:schedule (utils/json-counted-vector
+                                   (timers/start (metrics/waiter-timer "service-launch-overhead" "schedule-time"))
+                                   (timers/start (metrics/service-timer service-id "service-launch-overhead" "schedule-time")))
+                       :total (utils/json-counted-vector
+                                (timers/start (metrics/waiter-timer "service-launch-overhead" "init-to-healthy-time"))
+                                (timers/start (metrics/service-timer service-id "service-launch-overhead" "init-to-healthy-time")))}))]
+    (map->ServiceInstancesLaunchTracker
+      {:service-id service-id
+       :known-instance-ids #{}
+       :scheduling-instance-timer-contexts (vec ctxs)
+       :starting-instance-context-maps []})))
+
+(defn update-instance-launch-trackers
```

Comment:
It would be nice to get rid of this, but then we need a way to rebuild track the `known-instances` set on the next scale-up event. I'm not convinced there's a simple way to do that correctly. I can open an issue to revisit this to see if it can be further optimized, but I'd like to keep the trackers around for now.